### PR TITLE
Support for domain-specific heuristics encoded as heuristic directives

### PR DESCRIPTION
**Pull request by rtaupe**
These changes implement support for domain-specific heuristics encoded declaratively as heuristic directives within the input program.

### CODE REVIEW

**Comment by rtaupe** [comment]
@AntoniusW this pull request is now ready for review.

File: src/main/java/at/ac/tuwien/kr/alpha/solver/ChoiceInfluenceManager.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+			influencers.put(disabler, choicePoint);
+		}
+	}
+	
+	void updateAssignmentEntry(int currentDecisionLevel, Assignment.Entry entry) {
+		updateAssignmentFromChoicePoint(currentDecisionLevel, entry, influencers.get(entry.getAtom()));
+	}
+
+	void updateAssignmentFromChoicePoint(int currentDecisionLevel, Assignment.Entry entry, ChoicePoint choicePoint) {
+		if (choicePoint != null) {
```
Comment (lorenzleutgeb) [inline src/main/java/at/ac/tuwien/kr/alpha/solver/ChoiceInfluenceManager.java]: `if (choicePoint != null) { return; }` would save some space to the left.

**Comment by rtaupe** [comment]
@lorenzleutgeb, thanks for raising awareness for this. I have tried to fixe some of the issues, please have a look.
But I´m not certain if all the rules really make sense ... e.g., are 4 arguments to a constructor really too many?

**Comment by lorenzleutgeb** [comment]
Does this PR provide an easy way to run experimentally check the impact of heuristics on Alpha's running time?

If yes, how? And if no, one possible implementation could be:

```
$ java -jar alpha.jar -i example.asp
# Now, just out of curiousity, try without heuristics:
$ java -jar apha.jar --no-heuristic-directives -i example.asp
# Compare times of the runs.
```

**Comment by rtaupe** [comment]
@lorenzleutgeb The heuristic directives can be ignored by setting the command-line switch `-ids` or `--ignoreDomSpecHeuristics`.

**Comment by lorenzleutgeb** [comment]
> @lorenzleutgeb The heuristic directives can be ignored by setting the command-line switch `-ids` or `--ignoreDomSpecHeuristics`.

Thank you. I didn't check thoroughly enough.

**Comment by AntoniusW** [comment]
There are several tests failing, because they take too much time. Looks like the culprit is the naive solver again. Maybe #146 is resolving those issues.

**Comment by rtaupe** [comment]
The reason why unit tests on this branch take more time than on master is that on master, `at.ac.tuwien.kr.alpha.solver.ChoiceInfluenceManager#checksEnabled` is never set to `true` and therefore `at.ac.tuwien.kr.alpha.solver.ChoiceInfluenceManager#checkActiveChoicePoint` is never executed. On this branch, it is executed and takes so much resources that many unit tests ran into time-out. With af72a7c8595938277be5bbc8473a31085f3cbddc, the frequency of checks is therefore reduced.

**Comment by rtaupe** [comment]
@AntoniusW since the tests are now passing, please continue your review.

**Comment by rtaupe** [comment]
Superseded by #213

**Pull request closed without merging**